lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs

```diff
@@ -0,0 +1,531 @@
+
+-- | Attempt to select a random entry from an empty set.
+--
+-- This should always return 'Nothing'.
+--
+prop_selectRandomSetMember_empty :: Property
+prop_selectRandomSetMember_empty = monadicIO $ do
+    result <- run $ UTxOIndex.selectRandomSetMember mempty
+    assert $ result == Nothing @()
+
+-- | Attempt to select a random entry from a singleton set with entry 'i'.
+--
+-- This should always return 'Just i'.
+--
+prop_selectRandomSetMember_singleton :: Int -> Property
+prop_selectRandomSetMember_singleton i = monadicIO $ do
+    result <- run $ UTxOIndex.selectRandomSetMember $ Set.singleton i
+    assert $ result == Just i
+
+-- | Verify that the random selection function chooses elements uniformly
+--   across the full range of a set.
+--
+prop_selectRandomSetMember_coversRangeUniformly :: Word8 -> Word8 -> Property
+prop_selectRandomSetMember_coversRangeUniformly i j =
+    withMaxSuccess 100_000 $ checkCoverageWith confidence $ monadicIO $ do
+        Just selected <- run $ UTxOIndex.selectRandomSetMember set
+        monitor $ cover 90 (elementCount > 1)
+            "set has more than 1 element"
+        monitor $ cover 50 (elementCount > 10)
+            "set has more than 10 elements"
+        monitor $ cover 10 (elementCount > 100)
+            "set has more than 100 elements"
+        monitor $ cover 5 (selected == smallest)
+            "selected smallest element"
+        monitor $ cover 5 (selected == greatest)
+            "selected greatest element"
+        monitor $ cover 50 (inLowerHalf selected)
+            "selected element in lower half of range"
+        monitor $ cover 50 (inUpperHalf selected)
+            "selected element in upper half of range"
+        monitor
+            $ counterexample
+            $ mconcat
+            $ fmap showKeyValuePair
+            [ ("smallest", show smallest)
+            , ("greatest", show greatest)
+            , ("midpoint", show midpoint)
+            , ("selected", show selected)
+            ]
+        assert $ fromIntegral smallest % 1 <= midpoint
+        assert $ fromIntegral greatest % 1 >= midpoint
+        assert $ selected `Set.member` set
+  where
+    confidence = stdConfidence {tolerance = 0.99}
+    elementCount = greatest - smallest
+    greatest = max i j
+    inLowerHalf r = fromIntegral r % 1 <= midpoint
+    inUpperHalf r = fromIntegral r % 1 >= midpoint
+    midpoint :: Rational
+    midpoint = (fromIntegral smallest + fromIntegral greatest) % 2
+    set = Set.fromList [smallest .. greatest]
+    showKeyValuePair (key, value) = key <> ": " <> value <> "\n"
+    smallest = min i j
+
+--------------------------------------------------------------------------------
+-- Utilities
+--------------------------------------------------------------------------------
+
+-- | Selects all UTxO entries matching a particular filter.
+--
+-- Returns a list of all the entries that matched, and an updated index with
+-- the selected entries removed.
+--
+selectAll
```

Comment:
I'm seriously wondering why this is not in `Cardano.Wallet.Primitive.Types.UTxOIndex.Internal`.

Reply:
> I'm seriously wondering why this is not in
> Cardano.Wallet.Primitive.Types.UTxOIndex.Internal.

There's really no use for it outside of tests.